// igniteui-angular/esm2015/lib/grids/grid/grid.details.pipe.js
import { Pipe } from '@angular/core';
import { GridBaseAPIService } from '../api.service';
/** @hidden */
export class IgxGridDetailsPipe {
    constructor(gridAPI) {
        this.gridAPI = gridAPI;
    }
    transform(collection, hasDetails, expansionStates, _pipeTrigger) {
        if (!hasDetails) {
            return collection;
        }
        const res = this.addDetailRows(collection, expansionStates);
        return res;
    }
    addDetailRows(collection, _expansionStates) {
        const result = [];
        collection.forEach((v) => {
            result.push(v);
            if (!this.gridAPI.grid.isGroupByRecord(v) && !this.gridAPI.grid.isSummaryRow(v) &&
                this.gridAPI.get_row_expansion_state(v)) {
                const detailsObj = { detailsData: v };
                result.push(detailsObj);
            }
        });
        return result;
    }
}
IgxGridDetailsPipe.decorators = [
    { type: Pipe, args: [{
                name: 'gridDetails',
                pure: true
            },] }
];
IgxGridDetailsPipe.ctorParameters = () => [
    { type: GridBaseAPIService }
];
