// igniteui-angular/esm2015/lib/tree/tree.service.js
import { Injectable } from '@angular/core';
/** @hidden @internal */
export class IgxTreeService {
    constructor() {
        this.expandedNodes = new Set();
        this.collapsingNodes = new Set();
        this.siblingComparer = (data, node) => node !== data && node.level === data.level;
    }
    /**
     * Adds the node to the `expandedNodes` set and fires the nodes change event
     *
     * @param node target node
     * @param uiTrigger is the event triggered by a ui interraction (so we know if we should animate)
     * @returns void
     */
    expand(node, uiTrigger) {
        var _a;
        this.collapsingNodes.delete(node);
        if (!this.expandedNodes.has(node)) {
            node.expandedChange.emit(true);
        }
        else {
            return;
        }
        this.expandedNodes.add(node);
        if (this.tree.singleBranchExpand) {
            (_a = this.tree.findNodes(node, this.siblingComparer)) === null || _a === void 0 ? void 0 : _a.forEach(e => {
                if (uiTrigger) {
                    e.collapse();
                }
                else {
                    e.expanded = false;
                }
            });
        }
    }
    /**
     * Adds a node to the `collapsing` collection
     *
     * @param node target node
     */
    collapsing(node) {
        this.collapsingNodes.add(node);
    }
    /**
     * Removes the node from the 'expandedNodes' set and emits the node's change event
     *
     * @param node target node
     * @returns void
     */
    collapse(node) {
        if (this.expandedNodes.has(node)) {
            node.expandedChange.emit(false);
        }
        this.collapsingNodes.delete(node);
        this.expandedNodes.delete(node);
    }
    isExpanded(node) {
        return this.expandedNodes.has(node);
    }
    register(tree) {
        this.tree = tree;
    }
}
IgxTreeService.decorators = [
    { type: Injectable }
];
