// igniteui-angular/esm2015/lib/tabs/bottom-nav/bottom-nav.component.js
import { Component } from '@angular/core';
import { IgxTabsBase } from '../tabs.base';
import { IgxTabsDirective } from '../tabs.directive';
/** @hidden */
let NEXT_BOTTOM_NAV_ITEM_ID = 0;
/**
 * Bottom Navigation component enables the user to navigate among a number of contents displayed in a single view.
 *
 * @igxModule IgxBottomNavModule
 *
 * @igxTheme igx-bottom-nav-theme
 *
 * @igxKeywords bottom navigation
 *
 * @igxGroup Layouts
 *
 * @remarks
 * The Ignite UI for Angular Bottom Navigation component enables the user to navigate among a number of contents
 * displayed in a single view. The navigation through the contents is accomplished with the tab buttons located at bottom.
 *
 * @example
 * ```html
 * <igx-bottom-nav>
 *     <igx-bottom-nav-item>
 *         <igx-bottom-nav-header>
 *             <igx-icon igxBottomNavHeaderIcon>folder</igx-icon>
 *             <span igxBottomNavHeaderLabel>Tab 1</span>
 *         </igx-bottom-nav-header>
 *         <igx-bottom-nav-content>
 *             Content 1
 *         </igx-bottom-nav-content>
 *     </igx-bottom-nav-item>
 *     ...
 * </igx-bottom-nav>
 * ```
 */
export class IgxBottomNavComponent extends IgxTabsDirective {
    constructor() {
        super(...arguments);
        /** @hidden */
        this._disableAnimation = true;
        /** @hidden */
        this.componentName = 'igx-bottom-nav';
    }
    /** @hidden */
    getNextTabId() {
        return NEXT_BOTTOM_NAV_ITEM_ID++;
    }
}
IgxBottomNavComponent.decorators = [
    { type: Component, args: [{
                selector: 'igx-bottom-nav',
                template: "<ng-container *ngFor=\"let tab of items; let i = index\">\r\n    <ng-container *ngTemplateOutlet=\"tab.panelTemplate\"></ng-container>\r\n</ng-container>\r\n\r\n<div\r\n    #tablist\r\n    role=\"tablist\"\r\n    class=\"igx-bottom-nav__menu igx-bottom-nav__menu--bottom\"\r\n    aria-orientation=\"horizontal\"\r\n>\r\n    <ng-container *ngFor=\"let tab of items; let i = index\">\r\n        <ng-container *ngTemplateOutlet=\"tab.headerTemplate\"></ng-container>\r\n    </ng-container>\r\n</div>\r\n",
                providers: [{ provide: IgxTabsBase, useExisting: IgxBottomNavComponent }]
            },] }
];
