// igniteui-angular/esm2015/lib/tabs/bottom-nav/bottom-nav-header.component.js
import { Component, HostBinding } from '@angular/core';
import { IgxTabHeaderDirective } from '../tab-header.directive';
import { IgxTabHeaderBase } from '../tabs.base';
export class IgxBottomNavHeaderComponent extends IgxTabHeaderDirective {
    /** @hidden */
    get cssClassSelected() {
        return this.tab.selected;
    }
    /** @hidden */
    get cssClassDisabled() {
        return this.tab.disabled;
    }
    /** @hidden */
    get cssClass() {
        return (!this.tab.disabled && !this.tab.selected);
    }
}
IgxBottomNavHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'igx-bottom-nav-header',
                template: "<ng-content></ng-content>\r\n",
                providers: [{ provide: IgxTabHeaderBase, useExisting: IgxBottomNavHeaderComponent }]
            },] }
];
IgxBottomNavHeaderComponent.propDecorators = {
    cssClassSelected: [{ type: HostBinding, args: ['class.igx-bottom-nav__menu-item--selected',] }],
    cssClassDisabled: [{ type: HostBinding, args: ['class.igx-bottom-nav__menu-item--disabled',] }],
    cssClass: [{ type: HostBinding, args: ['class.igx-bottom-nav__menu-item',] }]
};
