// igniteui-angular/esm2015/lib/tabs/tab-content.directive.js
import { Directive, ElementRef, HostBinding } from '@angular/core';
import { IgxTabItemDirective } from './tab-item.directive';
export class IgxTabContentDirective {
    /** @hidden */
    constructor(tab, elementRef) {
        this.tab = tab;
        this.elementRef = elementRef;
        /** @hidden */
        this.role = 'tabpanel';
    }
    /** @hidden */
    get tabIndex() {
        return this.tab.selected ? 0 : -1;
    }
    /** @hidden */
    get zIndex() {
        return this.tab.selected ? 'auto' : -1;
    }
    /** @hidden */
    get nativeElement() {
        return this.elementRef.nativeElement;
    }
    ;
}
IgxTabContentDirective.decorators = [
    { type: Directive }
];
IgxTabContentDirective.ctorParameters = () => [
    { type: IgxTabItemDirective },
    { type: ElementRef }
];
IgxTabContentDirective.propDecorators = {
    role: [{ type: HostBinding, args: ['attr.role',] }],
    tabIndex: [{ type: HostBinding, args: ['attr.tabindex',] }],
    zIndex: [{ type: HostBinding, args: ['style.z-index',] }]
};
