// igniteui-angular/esm2015/lib/tabs/tab-header.directive.js
import { Directive, ElementRef, HostBinding, HostListener } from '@angular/core';
import { PlatformUtil } from '../core/utils';
import { IgxTabItemDirective } from './tab-item.directive';
import { IgxTabsBase } from './tabs.base';
export class IgxTabHeaderDirective {
    /** @hidden */
    constructor(tabs, tab, elementRef, platform) {
        this.tabs = tabs;
        this.tab = tab;
        this.elementRef = elementRef;
        this.platform = platform;
        /** @hidden */
        this.role = 'tab';
    }
    /** @hidden */
    get tabIndex() {
        return this.tab.selected ? 0 : -1;
    }
    /** @hidden */
    get ariaSelected() {
        return this.tab.selected;
    }
    /** @hidden */
    get ariaDisabled() {
        return this.tab.disabled;
    }
    /** @hidden */
    onClick() {
        if (this.tab.panelComponent) {
            this.tabs.selectTab(this.tab, true);
        }
    }
    /** @hidden */
    get nativeElement() {
        return this.elementRef.nativeElement;
    }
    ;
}
IgxTabHeaderDirective.decorators = [
    { type: Directive }
];
IgxTabHeaderDirective.ctorParameters = () => [
    { type: IgxTabsBase },
    { type: IgxTabItemDirective },
    { type: ElementRef },
    { type: PlatformUtil }
];
IgxTabHeaderDirective.propDecorators = {
    role: [{ type: HostBinding, args: ['attr.role',] }],
    tabIndex: [{ type: HostBinding, args: ['attr.tabindex',] }],
    ariaSelected: [{ type: HostBinding, args: ['attr.aria-selected',] }],
    ariaDisabled: [{ type: HostBinding, args: ['attr.aria-disabled',] }],
    onClick: [{ type: HostListener, args: ['click',] }]
};
